packages/flutter_tools/lib/src/base/task_queue.dart

```diff
@@ -0,0 +1,98 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:collection';
+
+import '../globals_null_migrated.dart' as globals;
+
+/// A closure type used by the [TaskQueue].
+typedef TaskQueueClosure<T> = Future<T> Function();
+
+/// A task queue of Futures to be completed in parallel, throttling
+/// the number of simultaneous tasks.
+///
+/// The tasks return results of type T.
+class TaskQueue<T> {
+  /// Creates a task queue with a maximum number of simultaneous jobs.
+  /// The [maxJobs] parameter defaults to the number of CPU cores on the
+  /// system.
+  TaskQueue({int? maxJobs})
+      : maxJobs = maxJobs ?? globals.platform.numberOfProcessors;
```

Comment:
Wouldn't the constraining factors here by network connections and/or disk i/o? In other words, might not bad things happen if I had a ridiculous number of cores and they're all writing to the disk at the same time?

Reply:
I haven't need any issues on my 72 core machine. Once the cache is warm (which happens non-parallel when we `pub get` the dummy pubspec earlier), I don't think there's a lot of disk i/o, just to load the pubspecs.  There is probably more network i/o to check for new versions, etc., but it's not horrible.